### PR TITLE
doc(*) copy pdk docs into godoc comments

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
Copy/paste from https://docs.konghq.com/1.4.x/pdk/ almost verbatim.

Various discrepancies were made evident, noted most of them with TODO comments.

### CODE REVIEW

File: request/request.go [hunk 1]
```diff
@@ -1,65 +1,148 @@
...
 }
 
+// kong.Request.GetHost() returns the host component of the request’s URL,
+// or the value of the “Host” header. The returned value is normalized
+// to lower-case form.
 func (r Request) GetHost() (host string, err error) {
 	return r.AskString(`kong.request.get_host`)
 }
 
+// kong.Request.GetPort() returns the port component of the request’s URL.
```
Comment (javierguerragiraldez) [inline request/request.go]: Lua docs mention that this returns a number, why is it a string here?

File: request/request.go [hunk 2]
```diff
@@ -1,65 +1,148 @@
...
+//
+// Whether this function considers X-Forwarded-Proto or not depends
+// on several Kong configuration parameters:
+//
+//   - trusted_ips
+//   - real_ip_header
+//   - real_ip_recursive
+//
+// Note: we do not currently offer support for Forwarded HTTP Extension (RFC 7239)
+// since it is not supported by ngx_http_realip_module.
```
Comment (javierguerragiraldez) [inline request/request.go]: same string / number divergence here

File: request/request.go [hunk 3]
```diff
@@ -1,65 +1,148 @@
...
 func (r Request) GetForwardedPort() (port string, err error) {
 	return r.AskString(`kong.request.get_forwarded_port`)
 }
 
+// kong.Request.GetHttpVersion() returns the HTTP version
+// used by the client in the request, returning values
+// such as "1"", "1.1", "2.0", or nil for unrecognized values.
```
Comment (javierguerragiraldez) [inline request/request.go]: same string / number divergence here.    and does it return an error on weird values?

File: request/request.go [hunk 4]
```diff
@@ -1,65 +1,148 @@
...
 
+// kong.Request.GetRawQuery() returns the query component of the request’s URL.
+// It is not normalized in any way (not even URL-decoding of special characters)
+// and does not include the leading ? character.
 func (r Request) GetRawQuery() (string, error) {
 	return r.AskString(`kong.request.get_raw_query`)
 }
 
+// kong.Request.GetQueryArg() returns the value of the specified argument,
+// obtained from the query arguments of the current request.
```
Comment (javierguerragiraldez) [inline request/request.go]: missing argument?

File: request/request.go [hunk 5]
```diff
@@ -1,65 +1,148 @@
...
+// if an argument was not given a value, or an array if an argument
+// was given in the query string multiple times. Keys and values are
+// unescaped according to URL-encoded escaping rules.
+//
+// Note that a query string `?foo&bar` translates to two boolean true arguments,
+// and ?foo=&bar= translates to two string arguments containing empty strings.
+//
+// The max_args argument specifies the maximum number of returned arguments.
+// Must be greater than 1 and not greater than 1000, or -1 to specify the
+// default limit of 100 arguments.
```
Comment (javierguerragiraldez) [inline request/request.go]: this return type is really uncomfortable to use on Go.  I think values should always be an array of strings:

	?foo&bar =>  {"foo": [], "bar": []}
	?foo=&bar= => {"foo": [""], "bar": [""]}

File: response/response.go [hunk 6]
```diff
@@ -28,30 +85,85 @@ func (r Response) GetHeaders(max_headers int) (res map[string]interface{}, err e
...
+//
 func (r Response) GetSource() (string, error) {
 	return r.AskString(`kong.response.get_source`)
 }
 
+// kong.Response.SetStatus() allows changing the downstream response
+// HTTP status code before sending it to the client.
+//
+// This function should be used in the header_filter phase,
+// as Kong is preparing headers to be sent back to the client.
```
Comment (javierguerragiraldez) [inline response/response.go]: external plugins don't get the `header_filter` phase.

File: response/response.go [hunk 7]
```diff
@@ -28,30 +85,85 @@ func (r Response) GetHeaders(max_headers int) (res map[string]interface{}, err e
...
 func (r Response) SetStatus(status int) error {
 	_, err := r.Ask(`kong.response.set_status`, status)
 	return err
 }
 
+// kong.Response.SetHeader() sets a response header with the given value.
+// This function overrides any existing header with the same name.
+//
+// This function should be used in the header_filter phase,
+// as Kong is preparing headers to be sent back to the client.
```
Comment (javierguerragiraldez) [inline response/response.go]: external plugins don't get the `header_filter` phase.

File: response/response.go [hunk 8]
```diff
@@ -28,30 +85,85 @@ func (r Response) GetHeaders(max_headers int) (res map[string]interface{}, err e
...
 
+// kong.Response.AddHeader() adds a response header with the given value.
+// Unlike kong.Response.SetHeader(), this function does not remove
+// any existing header with the same name. Instead, another header
+// with the same name will be added to the response. If no header
+// with this name already exists on the response, then it is added
+// with the given value, similarly to kong.Response.SetHeader().
+//
+// This function should be used in the header_filter phase,
+// as Kong is preparing headers to be sent back to the client.
```
Comment (javierguerragiraldez) [inline response/response.go]: external plugins don't get the `header_filter` phase.

File: response/response.go [hunk 9]
```diff
@@ -28,30 +85,85 @@ func (r Response) GetHeaders(max_headers int) (res map[string]interface{}, err e
...
 func (r Response) AddHeader(k string, v string) error {
 	_, err := r.Ask(`kong.response.add_header`, k, v)
 	return err
 }
 
+// kong.Response.ClearHeader() removes all occurrences of the specified header
+// in the response sent to the client.
+//
+// This function should be used in the header_filter phase,
+// as Kong is preparing headers to be sent back to the client.
```
Comment (javierguerragiraldez) [inline response/response.go]: external plugins don't get the `header_filter` phase.

File: response/response.go [hunk 10]
```diff
@@ -28,30 +85,85 @@ func (r Response) GetHeaders(max_headers int) (res map[string]interface{}, err e
...
 	return err
 }
 
+// kong.Response.SetHeaders() sets the headers for the response.
+// Unlike kong.Response.SetHeader(), the headers argument must be a map
+// in which each key is a string (corresponding to a header’s name),
+// and each value is a string, or an array of strings.
+//
+// This function should be used in the header_filter phase,
+// as Kong is preparing headers to be sent back to the client.
```
Comment (javierguerragiraldez) [inline response/response.go]: external plugins don't get the `header_filter` phase.

File: service/request/request.go [hunk 11]
```diff
@@ -1,62 +1,131 @@
...
 	_, err := r.Ask(`kong.service.request.set_method`, method)
 	return err
 }
 
+// kong.ServiceRequest.SetQuery() sets the querystring of the request to the Service.
+//
+// Unlike kong.ServiceRequest.SetRawQuery(), the query argument must be a table
+// in which each key is a string (corresponding to an arguments name), and each
+// value is either a boolean, a string or an array of strings or booleans.
+// Additionally, all string values will be URL-encoded.
```
Comment (javierguerragiraldez) [inline service/request/request.go]: this function takes a string argument, not a map.

File: service/request/request.go [hunk 12]
```diff
@@ -1,62 +1,131 @@
...
 }
 
+// kong.ServiceRequest SetRawBody() sets the body of the request to the Service.
+//
+// The body argument must be a string and will not be processed in any way.
+// This function also sets the Content-Length header appropriately.
+// To set an empty body, one can give an empty string "" to this function.
+//
+// For a higher-level function to set the body based on the request content type,
+// see kong.ServiceRequest.SetBody().
```
Comment (javierguerragiraldez) [inline service/request/request.go]: there's no `kong.ServiceRequest.SetBody()` ... yet?

File: request/request.go (hunk 1 above)
Comment (hishamhm) [reply request/request.go]: Looks like it's just a bug in the go-pdk? Is there a `AskNumber` function?

**Comment by javierguerragiraldez** [comment]
subsumed into #18

**Pull request closed without merging**